### PR TITLE
Add RCon support

### PR DESCRIPTION
**Pull request by Aaron1011**
This adds [RCon](http://wiki.vg/Rcon) events, as well as an `RConSource`, to the API.

Notable components include:

* An `RConSource`, which provides the `InetSocketAddress` used by an `RCon` client.
* Serveral `RCon`-related events:
 * The base `RConEvent`, which provides the client-supplied ID and payload, as well as the initiating `RConSource`.
* `RConLoginEvent`, which is fired when an `RCon` client authenticates.
* `RConCommandEvent`, which is fired when an `RCon` client executes a command. Normally, commands won't be executed unless the client has successfully authenticated. However, an option is provided to override this.

Some additional notes:

* The request ID is specified with each packet - thus, it is not available from an instance of `RConSource`.
* Access to the `RCON` password set in `server.properties` is made available through the API. While I feel that this is useful for cases such as allowing multiple correct passwords, @Deamon5550 has raised concerns about malicious plugins taking advantage of this.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/rcon/RConCommandEvent.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * <p>Normally, if an {@link RConSource} is not {@link RConSource#getLoggedIn()} logged in},
+     * the supplied {@link #getRawPayload() command} will not be executed.
+     *
+     * The opposite scenario, preventing the execution of a command by a logged in
+     * client, can be achieved with {@link #setCancelled(boolean)}.</p>
+     *
+     * @return Whether to override the client's logged in status
+     */
+    boolean overrideLoggedIn();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/event/rcon/RConCommandEvent.java]: I'd appreciate any suggestions for improving the name of this method and the corresponding setter.

**Comment by kenzierocks** [comment]
FYI, it's `RCON`, not `Rcon`. Sources: [[1]](http://wiki.vg/Rcon) [[2]](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-tools/1262672-minecraft-server-rcon-query-utility)

File: src/main/java/org/spongepowered/api/event/rcon/RConCommandEvent.java (hunk 1 above)
Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/event/rcon/RConCommandEvent.java]: `ignoreLoginStatus()`?

File: src/main/java/org/spongepowered/api/event/rcon/RConEvent.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.rcon;
+
+import org.spongepowered.api.util.command.source.RConSource;
+
+public interface RConEvent {
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/event/rcon/RConEvent.java]: extends Event?

**Comment by Minecrell** [comment]
@kenzierocks Although you're right as per Google style it's `Rcon`, not `RCON`. (https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s5.3-camel-case)

File: src/main/java/org/spongepowered/api/event/rcon/RCONEvent.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+     * Gets the ID provided by the client for this request
+     *
+     * <p>The provided ID is sent back in the response to the client.
+     *
+     * If authorization failed (password is incorrect, or client didn't authenticate),
+     * the ID will be set to -1 in the response.</p>
+     *
+     * @return The ID provided by the client
+     */
+    int getID();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/rcon/RCONEvent.java]: Should be `getId()`

File: src/main/java/org/spongepowered/api/Server.java [hunk 4]
```diff
@@ -206,6 +206,25 @@
...
 
     /**
+     * Tests whether the server has RCON enabled.
+     *
+     * <p>RCON allows commands to be remotely run on the server, after RCON
+     * client have authenticated with a password.</p>
+     *
+     * @return True if enabled, false if not
+     */
+    boolean getRCONEnabled();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/Server.java]: `isRconEnabled()`?

**Comment by kenzierocks** [comment]
@minecrell Yes, it was more for the javadocs.

File: src/main/java/org/spongepowered/api/event/rcon/RCONLoginEvent.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Gets the password provided by the client.
+     *
+     * @return The password provided by the client
+     */
+    String getPassword();
+
+    @Override
+    String getRawPayload();
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/event/rcon/RCONLoginEvent.java]: Why overwriting it here?
Has anything changed the method?

File: src/main/java/org/spongepowered/api/util/command/source/RCONSource.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     * @return Whether this client is logged in
+     */
+    boolean getLoggedIn();
+
+    /**
+     * Sets whether this client is logged in, or authenticated.
+     *
+     * @param loggedIn Whether this client is logged in
+     */
+    void setLoggedIn(boolean loggedIn);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/util/command/source/RCONSource.java]: When would you need to set this?

Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/util/command/source/RCONSource.java]: This allows for plugins to do things like per-IP passwords, or allow multiple correct passwords.

**Comment by Minecrell** [comment]
@Aaron1011 The Travis build failed :3

**Comment by sk89q** [comment]
Arguably the RCON stuff should be on some sort of RconProvider (name pending), which, even better, was a service, but I'm open to discussion.

File: src/main/java/org/spongepowered/api/event/rcon/RconCommandEvent.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    boolean isIgnoringLoginStatus();
+
+    /**
+     * Sets whether {@link RconSource#getLoggedIn()} will be ignored when
+     * determining whether to execute the supplied {@link #getRawPayload() command}.
+     *
+     * <p>Normally, if an {@link RconSource} is not {@link RconSource#getLoggedIn()} logged in},
+     * the supplied {@link #getRawPayload() command} will not be executed.
+     *
+     * The opposite scenario, preventing the execution of a command by a logged in
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/rcon/RconCommandEvent.java]: Missing `<p>`

File: src/main/java/org/spongepowered/api/event/rcon/RconCommandEvent.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    RconSource getSource();
+
+    /**
+     * Gets whether {@link RconSource#getLoggedIn()} will be ignored when
+     * determining whether to execute the supplied {@link #getRawPayload() command}.
+     *
+     * <p>Normally, if an {@link RconSource} is not {@link RconSource#getLoggedIn()} logged in},
+     * the supplied {@link #getRawPayload() command} will not be executed.
+     *
+     * The opposite scenario, preventing the execution of a command by a logged in
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/rcon/RconCommandEvent.java]: Missing `<p>`

File: src/main/java/org/spongepowered/api/event/rcon/RconEvent.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import org.spongepowered.api.util.command.source.RconSource;
+
+public interface RconEvent extends GameEvent {
+
+    /**
+     * Gets the ID provided by the client for this request
+     *
+     * <p>The provided ID is sent back in the response to the client.
+     *
+     * If authorization failed (password is incorrect, or client didn't authenticate),
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/rcon/RconEvent.java]: Put this on a new paragraph or remove the empty lines

**Pull request closed without merging**